[Core] Rework getSpeed()

Xaver-DaRed opened the pull request:
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Tries to get us closer to matching what retail does with speed in the following ways:
- Creates a set of modifiers to separate movement speed sources.
- Makes possitive movement speed bonuses from gear NOT stack with each other and take the highest only instead.
- Makes Mazurka and quickening not stack and take the highest of the two.
- Makes weight and curse, among others, stack with each other and be multiplicative to other sources.
- Fixes an issue with !speed GM command when using movement gear.

Resolves #2482, resolves #2499, resolves #2501 

## Retail observations:
- Positive movement speed from gear DOES NOT stack with each other. Only highest applies.
- Negative movement speed from gear DOES stack with each other.
- Flee puts you at 80 speed. 2 pieces of dusk gear did NOT affect movement speed. The third one did. We concluded flee gives a bonus higher than the cap.
- Max speed is 80. There is no possible way that we could come up a different combination which allowed us to surpass this limit.
  - Combination -> +18% speed ring + Chocobo mazurka + sprinter shoes useable item effect + PERFECT Bolters roll + KI resulted in flee speed. 
- Positive movement speed from gear resulted in diminishing returns when combined with negative gear.
  - Bloody pants + dusk mask should result in ABOUT the same speed as regular speed. However, it was notizably slower, just not THAT much.
- Weight effects are concluded to be multiplicative. Full dusk gear + mollusca mantle + weight + curse resulted in a very slow movement speed, but not much different than just gear. If it was additive, we would have just stoped.
  - This was not the case of a lower limit cap. Beadeaux curse + weight is slower than just curse. Similarly to positive movement speed from gear, we saw diminishing returns (in this case, a diminishing penalty, since just curse with no negative gear is still very notizable)

## Steps to test these changes

Go to retail and check the actions described above.
Perform those same actions here and see if its at the very least SOMEWHAT close.

Xaver-DaRed committed:
Rework getSpeed() and add new modifiers.

TeoTwawki commented:
> Max speed is 80. There is no possible way that we could come up a different combination which allowed us to surpass this limit.

status effects can exceed this. stronger versions of quickening can be stacked up (kupower maxed from resting) and **feast of swords in modern puts you at 200**)

so its not a hard cap on overall speed. its more like the way we handle haste where each source has a cap and dual wield and delay reductions and then they all get shmushed together to create your attack speed.

TeoTwawki commented on src/map/entities/battleentity.cpp:
```diff
@@ -226,17 +226,44 @@ int32 CBattleEntity::GetMaxMP() const
+        baseSpeed   = 40 + settings::get<int8>("map.MOUNT_SPEED_MOD");
+        outputSpeed = baseSpeed * (100 + getMod(Mod::MOUNT_MOVE)) / 100;
+
+        return std::clamp<uint8>(outputSpeed, 0, 80);
```
please don't clamp final chocobo/mount to 80. We want the setting to allow ridiculous chocobo speeds. It won't get that high without someone intentionally putting the setting way up there.

TeoTwawki commented on scripts/enum/mod.lua:
```diff
@@ -912,6 +912,12 @@ xi.mod =
 
     STEP_TP_CONSUMED        = 1077, -- Modifies the amount of TP consumed by dancer steps
 
+    MOVE_SPEED_STACKABLE      = 75, -- Gear movement speed penalties, flee bonus, etc.
```
since these are all together they can be placed earlier in the list instead of at the end

TeoTwawki commented on src/map/entities/battleentity.cpp:
```diff
@@ -226,17 +226,44 @@ int32 CBattleEntity::GetMaxMP() const
+    outputSpeed = static_cast<uint8>(modifiedSpeed);
 
-    return std::clamp<uint8>(outputSpeed, std::numeric_limits<uint8>::min(), std::numeric_limits<uint8>::max());
+    return std::clamp<uint8>(outputSpeed, 0, 80);
```
same here about clamping to 80. Due to settings, you want to take the setting into account not just use the 80 retail has

TeoTwawki replied:
side thought: we really at some point need to map the flag bits that are packed with the speed value. it was being capped at uint8 max before (255) but it actually stops before that because of the non-speed bits stealing some space. you can witness the speed "rolling over" in the client by using the GM command to go "too high" and finding yourself in slow motion past 240

TeoTwawki commented on src/map/modifier.h:
```diff
@@ -980,13 +980,19 @@ enum class Mod
     DIG_BYPASS_FATIGUE = 1074, // Chocobo digging modifier found in "Blue Race Silks". Modifier works as a direct percent.
     BREATH_DMG_DEALT   = 1075, // Breath damage dealt
 
+    // New movement speed modifiers.
```
maybe just throw them right under where `move` is waiting to get deleted and before `mount_move` sits

MowFord commented on src/map/entities/battleentity.cpp:
```diff
@@ -226,17 +226,44 @@ int32 CBattleEntity::GetMaxMP() const
+        return std::clamp<uint8>(outputSpeed, 0, 80);
+    }
+
+    // Flee, KIs, Gear penalties, Bolters Roll.
```
so bolter's roll fits interestingly into this:

> Stacks multiplicatively with [Movement Speed](https://www.bg-wiki.com/ffxi/Movement_Speed) equipment and additively with either [Mazurka](https://www.bg-wiki.com/ffxi/Mazurka) or [Quickening](https://www.bg-wiki.com/ffxi/Quickening) (but not both).

As usual I've done no retail testing, just wanted to add my 2cents that according to this wording it appears that bolter's roll does need to just be part of the additiveMods here. I.E. the logic LGTM

Xaver-DaRed replied:
Yeah, I was happy to see that it "fitted nicely" so to say.

TeoTwawki commented on settings/default/map.lua:
```diff
@@ -104,6 +104,10 @@ xi.settings.map =
     MOB_SPEED_MOD = 0,
 
+    -- Modifier that limits how fast players can move. Bypassed by feast of swords. This does not prevent cheats, btw.
+    SPEED_CAP       = 80,
```
I don't believe this needs a setting, nor that the final speed value needs clamped to 80. By definition when we adjust the existing speed settings value we're already going non-retail, we just need to include the settings value for the final results max value (and also stay within uint8 bounds). We shouldn't be trying to decide the speed limit there, only that we're not rolling into an overflow/underflow.

Xaver-DaRed committed:
Edit item_mods to use the new speed modifiers

Co-authored-by: N3ckB3ard <N3ckB3ard@users.noreply.github.com>

Xaver-DaRed committed:
Add new "weight" modifier to status effects

Xaver-DaRed committed:
Change mob files to use new movement modifier

Xaver-DaRed committed:
Use new speed modifier in item files

Xaver-DaRed committed:
Finish converting old speed modifier

Xaver-DaRed committed:
GM speed bypass

- Repurpose old movement modifier for GM only speed mod
- Add cap bypass for GM speed
- Correct math used.

N3ckB3ard commented:
Linked to Issues:
#2501
#2499

cocosolos commented:
Just gonna leave this here. Maybe relevant for future research.
https://www.youtube.com/watch?v=qBBSZnut42Q

Xaver-DaRed commented:
I saw that video a few days ago.
I think this was fixed in retail, but if this wasnt fixed, I refuse to emulate it.

Thats why the clamps are there, so it can never underflow

TeoTwawki commented:
p.s. technically, #2501 is an infodump in discussions rather than an issue to close. But this does take care of 90% of what it lays out as being wrong so ima have to update it after this merges :)
